### PR TITLE
chore: upgrade jest (and related packages) to v29

### PR DESCRIPTION
**Pull request by flootr**
Related to #78711

## Proposed Changes

PR upgrades Jest (and related packages where necessary) to v29. We're currently on v27 which means we're jumping two major versions.

## Testing Instructions

Verify all tests are green

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by noahtallen** [comment]
Migrated ETK and Odyssey stats to the shared testing config here: #79097

**Comment by tyxla** [comment]
> Migrated ETK and Odyssey stats to the shared testing config here: #79097

Looks great and works well in my testing. Looking forward to it unblocking this one 🚀

**Comment by noahtallen** [comment]
Hey @flootr, I rebased this after merging #79097. I ran into some weird conflicts, so I ultimately removed a couple commits and did another dedupe after running `yarn up jest babel-jest 'jest-*'` again.

While I was rebasing, I also forgot for a moment that I was rebasing and ended up finding a simpler fix for the e2e config type issue 😅 

I also don't think we need `@types/jest` anymore -- npm reports that jest bundles it's own types now (https://www.npmjs.com/package/jest), so I removed it. I also removed `jest` itself from devDependencies in a handful of Calypso apps which don't use it

**Comment by tyxla** [comment]
> Hey @flootr, I rebased this after merging #79097. I ran into some weird conflicts, so I ultimately removed a couple commits and did another dedupe after running `yarn up jest babel-jest 'jest-*'` again.
> 
> While I was rebasing, I also forgot for a moment that I was rebasing and ended up finding a simpler fix for the e2e config type issue 😅
> 
> I also don't think we need `@types/jest` anymore -- npm reports that jest bundles it's own types now (https://www.npmjs.com/package/jest), so I removed it. I also removed `jest` itself from devDependencies in a handful of Calypso apps which don't use it

Neat! Aside from this needing a rebase, I guess it should also wait for https://github.com/Automattic/wp-calypso/pull/79176 to land since #79097 got reverted. 

For the protocol, #79176 is approved and can be merged already.

**Comment by noahtallen** [comment]
> Aside from this needing a rebase

done and marked as ready for final review! 🎉

File: client/state/data-layer/wpcom/gravatar-upload/test/index.js [hunk 1]
```diff
@@ -54,10 +54,6 @@ describe( '#announceSuccess()', () => {
 			},
 		} ) );
 	} );
-	afterAll( () => {
```
Comment (noahtallen) [inline client/state/data-layer/wpcom/gravatar-upload/test/index.js:57]: Do we still need to restore the previous FormData implementation?

**Review by noahtallen — APPROVED** [review]
Looks good to me!

I noticed this line of code last week; we might be able to remove it:

https://github.com/Automattic/wp-calypso/blob/1043c246e944255bfdd85100a0f0d8c9c8dfc739/test/client/setup-test-framework.js#L43-L48

**Comment by tyxla** [comment]
> I noticed this line of code last week; we might be able to remove it:

Good spot, removed.

File: client/state/data-layer/wpcom/gravatar-upload/test/index.js (hunk 1 above)
Comment (tyxla) [reply client/state/data-layer/wpcom/gravatar-upload/test/index.js:57]: I don't think we need it; just removed it.

**Review by tyxla — APPROVED** [review]
Works well for me and the codechanges make sense! Nice work y'all!

LGTM 🚢

**Comment by tyxla** [comment]
I'll proceed to ship this one.

**Pull request merged**